napps/kytos/of_core/main.py

```diff
@@ -107,31 +109,100 @@ def handle_features_reply(self, event):
 
         switch.update_features(features)
 
-    @listen_to('kytos/core.messages.openflow.new')
-    def handle_new_openflow_message(self, event):
+        if (event.source.state == event.source.STATES.SETUP and
+                event.source.protocol.state == 'waiting_features_reply'):
+            event.source.protocol.state = 'handshake_complete'
+            event.source.state = event.source.STATES.ESTABLISHED
+            log.info('Handshake complete on connection: %s', event.source.id)
+            event_raw = KytosEvent(name='kytos/of_core.handshake_complete',
+                                   content={'source': event.source})
+            self.controller.buffers.app.put(event_raw)
+
+    @staticmethod
+    def of_slicer(data):
+        data_len = len(data)
+        pkts = []
+        while data_len > 3:
+            length_field = struct.unpack('!H', data[2:4])[0]
+            if data_len >= length_field:
+                pkts.append(data[:length_field])
+                data = data[length_field:]
+                data_len = len(data)
+        return pkts, data
+
+    @listen_to('kytos/core.server.data.in')
+    def handle_new_server_data(self, event):
         """Handle a RawEvent and generate a kytos/core.messages.in.* event.
 
         Args:
             event (KytosEvent): RawEvent with openflow message to be unpacked
         """
-        log.debug('RawOpenFlowMessage received by RawOFMessage handler')
-
-        # creates an empty OpenFlow Message based on the message_type defined
-        # on the unpacked header.
-        message = new_message_from_header(event.content['header'])
-        binary_data = event.content['binary_data']
-
-        # The unpack will happen only to those messages with body beyond header
-        if binary_data and len(binary_data) > 0:
-            message.unpack(binary_data)
-        log.debug('RawOpenFlowMessage unpacked')
+        packets, event.source.remaining_data = \
+            self.of_slicer(event.source.remaining_data +
+                           event.content['new_data'])
+
+        # if it is a new connection, we are expecting a hello packet
+        if event.source.state == event.source.STATES.NEW:
+            try:
```

Comment:
If you change the `if clause` from line 145 to:
`if packets and event.source.state == CONNECTION_STATE.NEW:`
you won't need this `try/except/else` here, right? Then you can just do:
```
hello_packet = packets.pop(0)
if self.validate_hello(hello_packet):
    ....
```

And also, you probably should add an `elif` clause for this `if`, like:
```
if packets and event.source.state == CONNECTION_STATE.NEW:
    ....
elif packets:
    for packet in packets:
        ....
```
Since the `packeting` processing (unpacking messages) should occur just after changing the connection state, or not?